spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/planning/RewrittenRowLevelCommand.scala

```diff
@@ -0,0 +1,112 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.spark.sql.catalyst.planning
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.plans.logical.ReplaceData
+import org.apache.spark.sql.catalyst.plans.logical.RowLevelCommand
+import org.apache.spark.sql.catalyst.plans.logical.UpdateTable
+import org.apache.spark.sql.connector.catalog.Table
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2ScanRelation
+
+/**
+ * An extractor for operations such as DELETE and MERGE that require rewriting data.
+ *
+ * This class extracts the following entities:
+ *  - the row-level command (such as DeleteFromIcebergTable);
+ *  - the scan relation in the rewrite plan that can be either DataSourceV2Relation or
+ *  DataSourceV2ScanRelation depending on whether the planning has already happened;
+ *  - the current rewrite plan.
+ */
+object RewrittenRowLevelCommand {
+  type ReturnType = (RowLevelCommand, LogicalPlan, LogicalPlan)
+
+  def unapply(plan: LogicalPlan): Option[ReturnType] = plan match {
+    case c: RowLevelCommand if c.rewritePlan.nonEmpty =>
+      val rewritePlan = c.rewritePlan.get
+
+      // both ReplaceData and WriteDelta reference a write relation
+      // but the corresponding scan relation should be at the bottom of the write plan
+      // both the write and scan relations will share the same RowLevelOperationTable object
+      // that's why it is safe to use reference equality to find the needed scan relation
+
+      val allowScanDuplication = c match {
+        // group-based updates that rely on the union approach may have multiple identical scans
+        case _: UpdateTable if rewritePlan.isInstanceOf[ReplaceData] => true
+        case _ => false
+      }
+
+      rewritePlan match {
+        case replaceData: ReplaceData =>
+          replaceData.table match {
+            case DataSourceV2Relation(table, _, _, _, _) =>
+              val scanRelation = findScanRelation(table, replaceData.query, allowScanDuplication)
+              scanRelation.map((c, _, replaceData))
+            case _ =>
+              None
+          }
+      }
+
+    case _ =>
+      None
+  }
+
+  private def findScanRelation(
```

Comment:
This is called `findScanRelation` but may return either a `DataSourceV2ScanRelation` or a relation that is not a `*ScanRelation`. Should this be `findRelationForTable` instead?

Reply:
I meant that the original relation in row-level commands is kind of split into "write" and "scan" relations for the same table. Then we read from the scan relation and write into the write relation.

I am still wondering what would be the best naming here. I want to refer to the relation we read from.

Reply:
One option can be to call them "read" and "write" relations to avoid using "scan".

Reply:
I like "read", but this is minor.

Reply:
I've updated to use "read".